### PR TITLE
WT-12897 Fix potential data race condition in fail_fs.c

### PR DESCRIPTION
**Pull request by jiechenbo**
It was not possible to hit the race condition before this fix. The `fail_fs->fail_enabled` would only change if the environment variable was changed underneath the caller of the function. Now the change ensures that fail_enabled, allow_reads are only read under a lock.

### CODE REVIEW

**Review by etienneptl — APPROVED** [review]

File: ext/test/fail_fs/fail_fs.c [hunk 1]
```diff
@@ -239,10 +239,12 @@ fail_file_read(
     } else
         read_ops = ++fail_fs->read_ops;
 
-    fail_fs_unlock(&fail_fs->lock);
-
-    if (fail_fs->fail_enabled && fail_fs->allow_reads != 0 && read_ops % fail_fs->allow_reads == 0)
+    if (fail_fs->fail_enabled && fail_fs->allow_reads != 0 &&
+      read_ops % fail_fs->allow_reads == 0) {
+        fail_fs_unlock(&fail_fs->lock);
```
Comment (marcbutler) [inline ext/test/fail_fs/fail_fs.c]: As discussed off line for both this function and "write", I'd suggest capturing the return code to `ret` and jumping (goto) to a common return path at the end of the function.

File: ext/test/fail_fs/fail_fs.c [hunk 2]
```diff
@@ -239,11 +239,13 @@ fail_file_read(
     } else
         read_ops = ++fail_fs->read_ops;
 
+    if (fail_fs->fail_enabled && fail_fs->allow_reads != 0 &&
+      read_ops % fail_fs->allow_reads == 0) {
+        ret = fail_fs_simulate_fail(fail_fh, session, read_ops, "read");
+        goto err;
```
Comment (etienneptl) [inline ext/test/fail_fs/fail_fs.c:245]:
```suggestion
        WT_ERR(fail_fs_simulate_fail(fail_fh, session, read_ops, "read"););
```

File: ext/test/fail_fs/fail_fs.c [hunk 3]
```diff
@@ -348,11 +354,12 @@ fail_file_write(
     } else
         write_ops = ++fail_fs->write_ops;
 
-    fail_fs_unlock(&fail_fs->lock);
-
     if (fail_fs->fail_enabled && fail_fs->allow_writes != 0 &&
-      write_ops % fail_fs->allow_writes == 0)
-        return (fail_fs_simulate_fail(fail_fh, session, write_ops, "write"));
+      write_ops % fail_fs->allow_writes == 0) {
+        ret = fail_fs_simulate_fail(fail_fh, session, write_ops, "write");
```
Comment (etienneptl) [inline ext/test/fail_fs/fail_fs.c:359]: ditto

File: ext/test/fail_fs/fail_fs.c (hunk 2 above)
Comment (jiechenbo) [reply ext/test/fail_fs/fail_fs.c:245]: The fail_fs.c file is outside of WiredTiger source code. The WT_ERR() macro is able to used here. I was following the file's other examples of err labels and they use goto err: explicitly. What do you think should be the correct style for this file? @etienneptl

Comment (etienneptl) [reply ext/test/fail_fs/fail_fs.c:245]: I don't think we should use `WT_ERR` if this is not internal code, good call.

**Review by etienneptl — APPROVED** [review]

**Comment by marcbutler** [comment]
Hi @ddanderson ,

While discussing this with @jiechenbo my reading of {{fail_file_write}} is that when it breaks the write into chunks:
```
/* Break writes larger than 1GB into 1GB chunks. */
```

It potentially introduces a race-condition where by different threads writing to the same file may incorrectly interleave calls to `pwrite`.

**Comment by ddanderson** [comment]
> Hi @ddanderson ,
> 
> While discussing this with @jiechenbo my reading of {{fail_file_write}} is that when it breaks the write into chunks:
> 
> ```
> /* Break writes larger than 1GB into 1GB chunks. */
> ```
> 
> It potentially introduces a race-condition where by different threads writing to the same file may incorrectly interleave calls to `pwrite`.

@marcbutler I think it is possible that writes are interleaved, but I don't think that's a problem.  Since we're using pwrite, not write, I think each thread is providing it's own offset into the file.  So while the writes may be somewhat concurrent, they are each going to the intended place on disk.  Maybe I'm missing something?

I think what might happen, that shouldn't happen with a "real" file system is that a write of N bytes to fail_fs might write less than N bytes, and if the system or application crashes, that the resulting write will be partial.  But I don't think that generally matters at least in WT data files because 1) WT shouldn't reference the written page from any valid checkpoint unless the file_fs->write returns, and it must return to be part of the checkpoint.  2) Even if that does happen, WT detects partial writes via checksums, so we'd at least detect it.  I think we're protected with log file writes as well, basically with checksums.

At the moment I _think_ that file_fs isn't used in conjunction of tests that involve crashes and attempts at restart.  But it might not always be true, so good to think through these things.

**Comment by marcbutler** [comment]
@ddanderson Thanks for clarifying. I still think it is a potential problem if writing to the same (or overlapping) file offsets are accidentally reordered: that is T1 should write before T2 but T2 completes first. I'll take an action to lodge a ticket, and this can be discussed and captured there.

**Review by marcbutler — APPROVED** [review]
@jiechenbo Thanks for your patience.

**Pull request merged**